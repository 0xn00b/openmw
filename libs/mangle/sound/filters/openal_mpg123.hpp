#ifndef MANGLE_MPG123_OPENAL_H
#define MANGLE_MPG123_OPENAL_H

#include "input_filter.hpp"
#include "../sources/mpg123_source.hpp"
#include "../outputs/openal_out.hpp"

namespace Mangle {
namespace Sound {

/// A InputFilter that adds mpg123 decoding to OpenAL. Only supports
/// MP3 files.
class OpenAL_Mpg123_Factory : public InputFilter
{
 public:
  OpenAL_Mpg123_Factory()
    {
      set(SoundFactoryPtr(new OpenAL_Factory),
          SampleSourceLoaderPtr(new Mpg123Loader));
    }
};

}}
#endif
