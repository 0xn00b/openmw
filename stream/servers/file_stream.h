#ifndef MANGLE_STREAM_FILESERVER_H
#define MANGLE_STREAM_FILESERVER_H

#include "std_stream.h"
#include <fstream>

namespace Mangle {
namespace Stream {

/** Very simple file input stream, based on std::ifstream
 */
class FileStream : public StdStream
{
  std::ifstream file;

 public:
  FileStream(const std::string &name)
    : StdStream(&file)
    {
      file.open(name, ios::binary);
    }
  ~FileStream() { file.close(); }
};

typedef boost::shared_ptr<FileStream> FileStreamPtr;

}} // namespaces
#endif
