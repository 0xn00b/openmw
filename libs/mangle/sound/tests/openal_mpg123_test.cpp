#include <iostream>
#include <exception>

#include "../../stream/servers/file_stream.hpp"
#include "../filters/openal_mpg123.hpp"

using namespace std;
using namespace Mangle::Stream;
using namespace Mangle::Sound;

OpenAL_Mpg123_Factory mg;

void play(const char* name, bool stream=false)
{
  // Only load streams if the backend supports it
  if(stream && !mg.canLoadStream)
    return;

  cout << "Playing " << name;
  if(stream) cout << " (from stream)";
  cout << "\n";

  SoundPtr snd;

  try
    {
      if(stream)
        snd = mg.load(StreamPtr(new FileStream(name)));
      else
        snd = mg.load(name);

      snd->setStreaming(true);
      snd->play();

      while(snd->isPlaying())
        {
          usleep(10000);
          if(mg.needsUpdate) mg.update();
        }
    }
  catch(exception &e)
    {
      cout << "  ERROR: " << e.what() << "\n";
    }
}

int main(int argc, char**argv)
{
  if(argc != 2)
    cout << "Please specify an MP3 file\n";
  else
    play(argv[1]);
  return 0;
}
