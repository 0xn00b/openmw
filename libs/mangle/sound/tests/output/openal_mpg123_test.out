Please specify an MP3 file
